linux/LibNl.c

```diff
@@ -0,0 +1,220 @@
+#include "linux/LibNl.h"
+
+#include <dlfcn.h>
+
+#include <linux/netlink.h>
+#include <linux/taskstats.h>
+
+#include <netlink/attr.h>
+#include <netlink/handlers.h>
+#include <netlink/msg.h>
+
+
+static void* libnlHandle;
+static void* libnlGenlHandle;
+
+static void (*sym_nl_close)(struct nl_sock*);
+static int (*sym_nl_connect)(struct nl_sock*, int);
+static int (*sym_nl_recvmsgs_default)(struct nl_sock*);
+static int (*sym_nl_send_sync)(struct nl_sock*, struct nl_msg*);
+static struct nl_sock* (*sym_nl_socket_alloc)(void);
+static void (*sym_nl_socket_free)(struct nl_sock*);
+static int (*sym_nl_socket_modify_cb)(struct nl_sock*, enum nl_cb_type, enum nl_cb_kind, nl_recvmsg_msg_cb_t, void*);
+static void* (*sym_nla_data)(const struct nlattr*);
+static struct nlattr* (*sym_nla_next)(const struct nlattr*, int*);
+static int (*sym_nla_put_u32)(struct nl_msg*, int, uint32_t);
+static struct nl_msg* (*sym_nlmsg_alloc)(void);
+static struct nlmsghdr* (*sym_nlmsg_hdr)(struct nl_msg*);
+static void (*sym_nlmsg_free)(struct nl_msg*);
+
+static int (*sym_genl_ctrl_resolve)(struct nl_sock*, const char*);
+static int (*sym_genlmsg_parse)(struct nlmsghdr*, int, struct nlattr**, int, const struct nla_policy*);
+static void* (*sym_genlmsg_put)(struct nl_msg*, uint32_t, uint32_t, int, int, int, uint8_t, uint8_t);
+
+
+static int load_libnl(void) {
+   if (libnlHandle && libnlGenlHandle)
+      return 0;
+
+   libnlHandle = dlopen("libnl-3.so", RTLD_LAZY);
+   if (!libnlHandle) {
+      libnlHandle = dlopen("libnl-3.so.200", RTLD_LAZY);
+      if (!libnlHandle) {
+         goto dlfailure;
+      }
+   }
+
+   libnlGenlHandle = dlopen("libnl-genl-3.so", RTLD_LAZY);
+   if (!libnlGenlHandle) {
+      libnlGenlHandle = dlopen("libnl-genl-3.so.200", RTLD_LAZY);
+      if (!libnlGenlHandle) {
+         goto dlfailure;
+      }
+   }
+
+   /* Clear any errors */
+   dlerror();
+
+   #define resolve(handle, symbolname) do {                              \
+      *(void **)(&sym_##symbolname) = dlsym(handle, #symbolname);        \
+      if (!sym_##symbolname || dlerror() != NULL) {                      \
+         goto dlfailure;                                                 \
+      }                                                                  \
+   } while(0)
+
+   resolve(libnlHandle, nl_close);
+   resolve(libnlHandle, nl_connect);
+   resolve(libnlHandle, nl_recvmsgs_default);
+   resolve(libnlHandle, nl_send_sync);
+   resolve(libnlHandle, nl_socket_alloc);
+   resolve(libnlHandle, nl_socket_free);
+   resolve(libnlHandle, nl_socket_modify_cb);
+   resolve(libnlHandle, nla_data);
+   resolve(libnlHandle, nla_next);
+   resolve(libnlHandle, nla_put_u32);
+   resolve(libnlHandle, nlmsg_alloc);
+   resolve(libnlHandle, nlmsg_free);
+   resolve(libnlHandle, nlmsg_hdr);
+
+   resolve(libnlGenlHandle, genl_ctrl_resolve);
+   resolve(libnlGenlHandle, genlmsg_parse);
+   resolve(libnlGenlHandle, genlmsg_put);
+
+   #undef resolve
+
+   return 0;
+
+dlfailure:
```

Comment:
I wonder if we should set all pointers to `NULL` on failure (and when free'ing the lib..